### ISSUE TITLE
CI/CD | Refactor How GitHub Actions are Run So They're Efficent

### ISSUE THREAD

**Issue by ReedClanton** [ReedClanton/shell_base#49]
## Description ##

#24 will setup this project's pipeline, however, no attention will be paid to efficiency or even best practices. This issue is to cover refactoring the pipeline so it's faster and more robust.

## **High Level** Solution Description ##

The primary purpose of this issue is to learn about CI/CD best practices and implement them. Having said that, the bellow are already known to be needed:
- Dockerize testing so all tests may be run on a single image that already has `shellspec`, all shells covered by testing, `Kcov`, and any other items that each job currently has to re-install on the GitHub runner each time.

## Alternative Solution(s) ##

None.

## Blocked By ##

- [ ] #24 

## Blocking ##

- [ ] #33 
- [ ] #37
- [ ] #42
- [ ] #43 

## Related Issue(s) ##

- None

## Additional Context ##

This story was created in order to break out some of the work from #24 into another issue so that issue isn't **so** big.
